cosmos_service/src/app.py

```diff
@@ -0,0 +1,145 @@
+        result = await proc.wait()
+        queue.task_done()
+        
+        if result == OOM_ERROR_EXIT_CODE:
+            await asyncio.sleep(OOM_SLEEP_TIME)
+            await queue.put((job_output_dir, job_id))
+
+@app.post("/process/", status_code=202)
+async def process_document(pdf: UploadFile = File(...)):
+    """
+    Accept a new PDF document for COSMOS processing. Saves the PDF to disk, 
+    then adds it to a queue for subsequent processing
+    """
+    job_id = str(uuid.uuid4())
+    if not pdf.file or not pdf.filename:
+        raise HTTPException(status_code=400, detail="Poorly constructed form upload")
+
+    # Need to make a non-temporary directory to store PDF and job output due to
+    # asynchronous processing and retrieval, must be cleaned up in separate job
+    job_output_dir = f"{tempfile.gettempdir()}/{job_id}"
+    os.mkdir(job_output_dir)
+    try:
+        with open(f"{job_output_dir}/{pdf.filename}", "wb") as f:
+            shutil.copyfileobj(pdf.file, f)
+    except Exception:
+        return {"message": ":("}
+    finally:
+        pdf.file.close()
+
+    # populate the job in its default state (not started)
+    with SessionLocal() as session:
+        session.add(CosmosSessionJob(job_id, job_output_dir))
+        session.commit()
+
+    await queue.put((job_output_dir, job_id))
+    
+    return {
+        "message": "PDF Processing in Background", 
+        "job_id": job_id, 
+        "status_endpoint": f"/process/{job_id}/status",
+        "result_endpoint": f"/process/{job_id}/result"
+    }
+
+@app.get("/process/{job_id}/status")
+def get_processing_status(job_id: str):
+    """
+    Return the current status of a given pdf in the COSMOS processing queue. If `job.is_completed`,
+    then the results of the processing can be retrieved from `/process/{job_id}/result`
+    """
+    with SessionLocal() as session:
+        job : CosmosSessionJob = session.get(CosmosSessionJob, job_id)
+        if not job:
+            raise HTTPException(status_code=404, detail="Job not found")
+        return { 
+            "job_started": job.is_started,
+            "job_completed": job.is_completed,
+            "time_in_queue": job.time_in_queue,
+            "time_processing": job.time_processing,
+            "error": job.error
+        }
+
+@app.get("/process/{job_id}/result")
+def get_processing_result(job_id: str):
+    """
+    Return the zip file containing the results of a completed COSMOS pipeline. Return status 400 if the 
+    job is in a not-complete state
+    """
+    with SessionLocal() as session:
+        job = session.get(CosmosSessionJob, job_id)
+        if not job:
+            raise HTTPException(status_code=404, detail="Job not found")
+        elif not job.is_completed:
+            raise HTTPException(status_code=400, detail="Job not finished")
+    output_file = f"{job.output_dir}/cosmos_output.zip"
+    return FileResponse(output_file)
```

Comment:
Can we provide a filename in this response? I like the idea of retaining the original name of the PDF (maybe in the `CosmosSessionJob`?) and having a returned output of `{pdf_basename}_COSMOS_results.zip`